<a href="https://colab.research.google.com/github/ferdinandrafols/IA_LLMs/blob/main/GSI073_aula0_redes_neurais.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

GSI073 - Tópicos Especiais de Inteligência Artificial

Definição dos dados

In [2]:
import torch; import sklearn; from torch import nn

# 1. Carregar dados
iris = sklearn.datasets.load_iris()
X = iris.data        # 4 features: sépalas e pétalas
y = (iris.target == 1).astype(float)  # 1 se Versicolor, 0 caso contrário

# 2. Preparar dados para pytorch
X = torch.tensor(X, dtype=torch.float32)
y = torch.tensor(y, dtype=torch.float32).view(-1, 1)

# Adicionar print para ver o resultado
print("Variável X (features):")
print(X)
print("\nVariável y (target):")
print(y)

Variável X (features):
tensor([[5.1000, 3.5000, 1.4000, 0.2000],
        [4.9000, 3.0000, 1.4000, 0.2000],
        [4.7000, 3.2000, 1.3000, 0.2000],
        [4.6000, 3.1000, 1.5000, 0.2000],
        [5.0000, 3.6000, 1.4000, 0.2000],
        [5.4000, 3.9000, 1.7000, 0.4000],
        [4.6000, 3.4000, 1.4000, 0.3000],
        [5.0000, 3.4000, 1.5000, 0.2000],
        [4.4000, 2.9000, 1.4000, 0.2000],
        [4.9000, 3.1000, 1.5000, 0.1000],
        [5.4000, 3.7000, 1.5000, 0.2000],
        [4.8000, 3.4000, 1.6000, 0.2000],
        [4.8000, 3.0000, 1.4000, 0.1000],
        [4.3000, 3.0000, 1.1000, 0.1000],
        [5.8000, 4.0000, 1.2000, 0.2000],
        [5.7000, 4.4000, 1.5000, 0.4000],
        [5.4000, 3.9000, 1.3000, 0.4000],
        [5.1000, 3.5000, 1.4000, 0.3000],
        [5.7000, 3.8000, 1.7000, 0.3000],
        [5.1000, 3.8000, 1.5000, 0.3000],
        [5.4000, 3.4000, 1.7000, 0.2000],
        [5.1000, 3.7000, 1.5000, 0.4000],
        [4.6000, 3.6000, 1.0000, 0.2000],
        [5.

Vamos interpretar o que você está vendo.

---

## 1. O que é o **X** (features)

Trecho do início de `X`:

```text
tensor([[5.1000, 3.5000, 1.4000, 0.2000],
        [4.9000, 3.0000, 1.4000, 0.2000],
        ...
```

Isso significa:

* `X` é um tensor de shape **[150, 4]** (150 linhas, 4 colunas).
* Cada **linha** = 1 flor (1 observação do dataset Iris).
* Cada **coluna** = 1 característica (*feature*):

  1. comprimento da sépala
  2. largura da sépala
  3. comprimento da pétala
  4. largura da pétala

Então, por exemplo:

```text
[5.1000, 3.5000, 1.4000, 0.2000]
```

→ é uma flor com:

* sépala 5.1 × 3.5
* pétala 1.4 × 0.2

As primeiras 50 linhas, como você vê pelos números pequenos de pétala (≈ 1.x), são da classe **Setosa**.

---

## 2. O que é o **y** (target binário)

Trecho do início de `y`:

```text
tensor([[0.],
        [0.],
        [0.],
        ...
        [0.],
        [0.],
        [0.],
        [1.],
        [1.],
        [1.],
        ...
        [1.],
        [1.],
        [0.],
        [0.],
        ...
```

Aqui está a mágica:

Você definiu:

```python
y = (iris.target == 1).astype(float)
```

Isso transforma o problema em **classificação binária**:

* `1.0` → a flor é **Versicolor** (classe 1 no Iris)
* `0.0` → a flor **não é Versicolor** (ou seja, é Setosa (0) ou Virginica (2))

O padrão do Iris é:

* amostras 0–49 → Setosa (classe 0) → viram `0.`
* amostras 50–99 → Versicolor (classe 1) → viram `1.`
* amostras 100–149 → Virginica (classe 2) → viram `0.`

E é exatamente isso que o seu `y` mostra:

* um bloco grande de **0.** no começo (Setosa) ✅
* um bloco de **1.** no meio (Versicolor) ✅
* um bloco de **0.** no final (Virginica) ✅

Ou seja: o *target* está correto para a tarefa

> “prever se a flor é Versicolor (1) ou não (0)”.

---

## 3. Estrutura está perfeita pra ML / PyTorch

Você agora tem:

* `X`: `torch.Size([150, 4])`, `dtype=torch.float32`
* `y`: `torch.Size([150, 1])`, `dtype=torch.float32`

Isso está **perfeito** para:

* dividir em treino/teste,
* jogar num modelo `nn.Linear(4, 1)` (regressão logística),
* treinar com `BCEWithLogitsLoss`.


In [4]:
# 3. Definir modelo
import torch.nn.functional as F
class RedeNeural(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim):
        super(RedeNeural, self).__init__()
        self.fc1 = nn.Linear(input_dim, hidden_dim)
        self.fc2 = nn.Linear(hidden_dim, output_dim)

    def forward(self, x):
        x = F.relu(self.fc1(x))
        x = self.fc2(x)
        return x

# Adicionar print para ver o resultado
print("Estrutura da Rede Neural:")
print(RedeNeural(input_dim=4, hidden_dim=8, output_dim=1))

Estrutura da Rede Neural:
RedeNeural(
  (fc1): Linear(in_features=4, out_features=8, bias=True)
  (fc2): Linear(in_features=8, out_features=1, bias=True)
)


# Task
Instanciar o modelo, definir função de perda e otimizador, dividir os dados, treinar o modelo e avaliar o modelo.

## Instanciar o modelo

### Subtask:
Criar uma instância da classe `RedeNeural` com as dimensões apropriadas.


**Reasoning**:
Instantiate the neural network model with the specified dimensions.



In [5]:
modelo = RedeNeural(input_dim=4, hidden_dim=8, output_dim=1)

# Adicionar print para ver o resultado
print("Instância do modelo:")
print(modelo)

Instância do modelo:
RedeNeural(
  (fc1): Linear(in_features=4, out_features=8, bias=True)
  (fc2): Linear(in_features=8, out_features=1, bias=True)
)


## Definir função de perda e otimizador

### Subtask:
Escolher uma função de perda adequada para classificação binária e um otimizador (como Adam ou SGD) para atualizar os pesos do modelo.


**Reasoning**:
Import the necessary loss function and optimizer, then instantiate them with the model parameters.



In [6]:
import torch.optim as optim

# 3. Definir função de perda e otimizador
criterion = nn.BCEWithLogitsLoss()
optimizer = optim.Adam(modelo.parameters(), lr=0.01)

# Adicionar prints para confirmar a criação
print("Função de perda:")
print(criterion)
print("\nOtimizador:")
print(optimizer)

Função de perda:
BCEWithLogitsLoss()

Otimizador:
Adam (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    capturable: False
    decoupled_weight_decay: False
    differentiable: False
    eps: 1e-08
    foreach: None
    fused: None
    lr: 0.01
    maximize: False
    weight_decay: 0
)


## Dividir os dados

### Subtask:
Separar os dados `X` e `y` em conjuntos de treino e teste para avaliar o desempenho do modelo.


**Reasoning**:
Import the necessary function for splitting data and then use it to create training and testing sets, printing their shapes to verify the split.



In [7]:
from sklearn.model_selection import train_test_split

# 4. Dividir dados em treino e teste
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Adicionar prints para verificar as formas
print("Forma de X_train:", X_train.shape)
print("Forma de X_test:", X_test.shape)
print("Forma de y_train:", y_train.shape)
print("Forma de y_test:", y_test.shape)

Forma de X_train: torch.Size([120, 4])
Forma de X_test: torch.Size([30, 4])
Forma de y_train: torch.Size([120, 1])
Forma de y_test: torch.Size([30, 1])


## Treinar o modelo

### Subtask:
Implementar o loop de treinamento, que inclui forward pass, cálculo da perda, backward pass e atualização dos pesos.


**Reasoning**:
Implement the training loop including forward pass, loss calculation, backward pass, and weight update for a defined number of epochs.



In [8]:
# 5. Loop de treinamento
epochs = 1000
for epoch in range(epochs):
    # Forward pass
    outputs = modelo(X_train)
    loss = criterion(outputs, y_train)

    # Backward pass e otimização
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    # Imprimir a perda a cada 100 épocas
    if (epoch + 1) % 100 == 0:
        print(f'Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f}')

Epoch [100/1000], Loss: 0.4082
Epoch [200/1000], Loss: 0.1322
Epoch [300/1000], Loss: 0.0755
Epoch [400/1000], Loss: 0.0612
Epoch [500/1000], Loss: 0.0553
Epoch [600/1000], Loss: 0.0520
Epoch [700/1000], Loss: 0.0503
Epoch [800/1000], Loss: 0.0492
Epoch [900/1000], Loss: 0.0485
Epoch [1000/1000], Loss: 0.0479


## Avaliar o modelo

### Subtask:
Calcular métricas de desempenho (como acurácia) nos dados de teste.


**Reasoning**:
Calculate the accuracy of the trained model on the test set by obtaining predictions, converting them to probabilities and then to predicted classes, and comparing them with the true labels.



In [9]:
# 6. Avaliar o modelo
modelo.eval()
with torch.no_grad():
    outputs_test = modelo(X_test)
    probabilities = torch.sigmoid(outputs_test)
    predicted_classes = (probabilities > 0.5).float()
    accuracy = (predicted_classes == y_test).float().mean()

print(f'Acurácia nos dados de teste: {accuracy.item():.4f}')

Acurácia nos dados de teste: 1.0000


## Summary:

### Data Analysis Key Findings

*   A `RedeNeural` model was successfully instantiated with 4 input features, 8 hidden neurons, and 1 output neuron.
*   The `BCEWithLogitsLoss` function was chosen as the loss function, suitable for binary classification, and the `Adam` optimizer with a learning rate of 0.01 was selected to update model weights.
*   The dataset was split into training (80%) and testing (20%) sets using `train_test_split`.
*   The model was trained for 1000 epochs, showing a decrease in loss from approximately 0.4082 at epoch 100 to 0.0479 at epoch 1000, indicating learning.
*   The model achieved an accuracy of 1.0000 on the test dataset, demonstrating perfect classification on this specific test set.

### Insights or Next Steps

*   The perfect accuracy on the test set suggests potential overfitting or a very simple dataset. Further validation with a larger, more diverse dataset or cross-validation is recommended to confirm generalization.
*   Explore different network architectures, hyperparameters (learning rate, number of epochs, hidden layer size), and regularization techniques to assess their impact on performance and robustness.
